x/staking/keeper/msg_server.go

```diff
@@ -210,3 +213,134 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
 		CompletionTime: completionTime,
 	}, nil
 }
+
+func getShareTokenDenom(validatorAddress string, epochNumber int64) string {
+	return validatorAddress + strconv.Itoa(int(epochNumber))
+}
+
+func (k msgServer) TokenizeShares(goCtx context.Context, msg *types.MsgTokenizeShares) (*types.MsgTokenizeSharesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	valAddr, valErr := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if valErr != nil {
+		return nil, valErr
+	}
+	validator, found := k.GetValidator(ctx, valAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	_ = validator
+
+	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddress, valAddr)
+	if !found {
+		return nil, types.ErrNoDelegatorForAddress
+	}
+
+	delegationAmount := validator.Tokens.ToDec().Mul(delegation.GetShares()).Quo(validator.DelegatorShares)
+	if msg.Amount.Amount.GT(sdk.Int(delegationAmount)) {
+		return nil, types.ErrNotEnoughDelegationShares
+	}
+
+	shareTokenDenom := getShareTokenDenom(msg.ValidatorAddress, k.epochKeeper.GetEpochNumber(ctx))
+	shareToken := sdk.NewCoin(shareTokenDenom, msg.Amount.Amount)
+
+	err = k.bankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.Coins{shareToken})
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, sdk.AccAddress(valAddr), sdk.Coins{shareToken})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = k.Unbond(ctx, delegatorAddress, valAddr, sdk.NewDec(msg.Amount.Amount.Int64()))
```

Comment:
We don't want to unbond coins here. Because that will reduce the validator's voting power.

 we need to find a mechanism here to convert the shares to tokens while keeping the validators tokens the same.
 
 What I think we need to do is have two kind of shares the delegation. shares and tokenized shares.